include/pybind11/pybind11.h

```diff
@@ -1342,13 +1342,18 @@ class module_ : public object {
         // GraalPy doesn't support PyModule_GetFilenameObject,
         // so getting by attribute (see PR #5584)
         handle this_module = m_ptr;
-        result.attr("__file__") = this_module.attr("__file__");
+        if (hasattr(this_module, "__file__")) {
+            result.attr("__file__") = this_module.attr("__file__");
+        }
 #else
         handle this_file = PyModule_GetFilenameObject(m_ptr);
-        if (!this_file) {
+        if (this_file) {
+            result.attr("__file__") = this_file;
+        } else if (PyErr_ExceptionMatches(PyExc_SystemError) != 0) {
```

Comment:
Ugh ... if we have to catch a `SystemError` here to accommodate submodules, I regret posting [this comment on PR #5584](https://github.com/pybind/pybind11/pull/5584/files#r2017618042).

I think it'll be best to get rid of this `#else` branch then.

Reply:
It should be extremely rare, this is only for embedded submodules, normally it works. I think it's probably fine to leave this way.

Reply:
Up to you, I'm OK if you want to keep it.

What still makes me feel uneasy: the risk of masking unexpected `SystemError`s.

Which is of course extremely low, but not zero. (Google experience: anything that can happen, does happen.)

Reply:
The function is part of the Stable API and just accesses a dictionary. My two cents: the current implementation is safe.

Reply:
Do you have a link to the implementation handy?

What's on my mind: Why does it raise `SystemError`, and not `AttributeError`?

I'd feel more cozy ignoring an `AttributeError`.

But as I said, I'm fine with this.

Reply:
The function explicitly sets the system error. I could not see any system errors in the source of the dictionary.

- https://github.com/python/cpython/blob/7dddb4e667b5eb76cbe11755051ec139b0f437a9/Objects/moduleobject.c#L644
- https://github.com/python/cpython/blob/7dddb4e667b5eb76cbe11755051ec139b0f437a9/Objects/dictobject.c#L2392

Reply:
Thanks!

> The function explicitly sets the system error.

Seems like an unfortunate choice, to have it in the same class of exceptions as e.g. out-of-memory, but 🤷

Reply:
```
>>> SystemError.__mro__
(<class 'SystemError'>, <class 'Exception'>, <class 'BaseException'>, <class 'object'>)
>>> MemoryError.__mro__
(<class 'MemoryError'>, <class 'Exception'>, <class 'BaseException'>, <class 'object'>)
```

https://docs.python.org/3/library/exceptions.html#SystemError

> Raised when the interpreter finds an internal error, but the situation does not look so serious to cause it to abandon all hope.

Doesn’t seem too bad, I think it’s for things like this, exceptions that shouldn’t make it up to the user.